From bbcf1ad4ad009087432c74a066d7741e2f06b56e Mon Sep 17 00:00:00 2001
From: Pradeep Reddy <pradeep.reddy@morsemicro.com>
Date: Sat, 21 Jun 2025 03:14:25 +1000
Subject: [PATCH] mac80211: add halow specific mesh optimizations

A collection of smaller patches to improve mesh performance in rate
limited network links such as Wi-Fi HaLow. These include tuning
adjustments for buffered frames, timeout increases to reduce control
frame traffic, increases to maximum discovery times and so on.
---
 net/mac80211/mesh.h         |   8 +-
 net/mac80211/mesh_hwmp.c    | 151 ++++++++++++++++++++++++++++++------
 net/mac80211/mesh_pathtbl.c |   6 ++
 net/mac80211/rx.c           |  14 +---
 net/wireless/mesh.c         |  12 +--
 5 files changed, 149 insertions(+), 42 deletions(-)

diff --git a/net/mac80211/mesh.h b/net/mac80211/mesh.h
index 2350bf695dbb7..bc0c23423cc04 100644
--- a/net/mac80211/mesh.h
+++ b/net/mac80211/mesh.h
@@ -229,7 +229,13 @@ struct mesh_rmc {
 #define MESH_MAX_MPATHS		1024
 
 /* Number of frames buffered per destination for unresolved destinations */
-#define MESH_FRAME_QUEUE_LEN	10
+#define MESH_FRAME_QUEUE_LEN	50
+
+/* Best metric for single hop HaLow link */
+#define BEST_SINGLE_HOP_METRIC_HALOW  253
+
+/* Max path discovery timeout in msecs */
+#define MESH_MAX_MPATH_DISCOVERY_TIMEOUT      2000
 
 /* Public interfaces */
 /* Various */
diff --git a/net/mac80211/mesh_hwmp.c b/net/mac80211/mesh_hwmp.c
index ba5aeeb0836ff..73eb1bb34d799 100644
--- a/net/mac80211/mesh_hwmp.c
+++ b/net/mac80211/mesh_hwmp.c
@@ -132,13 +132,13 @@ static int mesh_path_sel_frame_tx(enum mpath_frame_type action, u8 flags,
 
 	switch (action) {
 	case MPATH_PREQ:
-		mhwmp_dbg(sdata, "sending PREQ to %pM\n", target);
+		mhwmp_dbg(sdata, "sending PREQ to %pM da:%pM\n", target, da);
 		ie_len = 37;
 		pos = skb_put(skb, 2 + ie_len);
 		*pos++ = WLAN_EID_PREQ;
 		break;
 	case MPATH_PREP:
-		mhwmp_dbg(sdata, "sending PREP to %pM\n", orig_addr);
+		mhwmp_dbg(sdata, "sending PREP to %pM da:%pM\n", orig_addr, da);
 		ie_len = 31;
 		pos = skb_put(skb, 2 + ie_len);
 		*pos++ = WLAN_EID_PREP;
@@ -380,6 +380,32 @@ u32 airtime_link_metric_get(struct ieee80211_local *local,
 	return (u32)result;
 }
 
+void ieee80211_mpath_dump(struct ieee80211_sub_if_data *sdata)
+{
+    struct mesh_path *mpath;
+    u32 exp_time = 0;
+    int idx = 0;
+
+    mhwmp_dbg(sdata, "************** mpath Table **************\n");
+    while (mpath = mesh_path_lookup_by_idx(sdata, idx)) {
+        if (time_before(jiffies, mpath->exp_time))
+            exp_time = jiffies_to_msecs(mpath->exp_time - jiffies);
+        else
+            exp_time = 0;
+
+        mhwmp_dbg(sdata, "idx:%d dst:%pM next_hop:%pM metric:%d hops:%d exp_time:%d flags:%x\n",
+            idx,
+            mpath->dst,
+            rcu_access_pointer(mpath->next_hop)->addr,
+            mpath->metric,
+            mpath->hop_count,
+            exp_time,
+            mpath->flags);
+
+        idx++;
+    }
+}
+
 /**
  * hwmp_route_info_get - Update routing info to originator and transmitter
  *
@@ -404,6 +430,7 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_local *local = sdata->local;
 	struct mesh_path *mpath;
 	struct sta_info *sta;
+	struct ieee80211_if_mesh *ifmsh = &sdata->u.mesh;
 	bool fresh_info;
 	const u8 *orig_addr, *ta;
 	u32 orig_sn, orig_metric;
@@ -412,6 +439,7 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 	bool flush_mpath = false;
 	bool process = true;
 	u8 hopcount;
+	bool new_path = false;
 
 	rcu_read_lock();
 	sta = sta_info_get(sdata, mgmt->sa);
@@ -463,22 +491,26 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 		mpath = mesh_path_lookup(sdata, orig_addr);
 		if (mpath) {
 			spin_lock_bh(&mpath->state_lock);
+			mhwmp_dbg(sdata, "%s from:%pM->%pM new_metric:%d metric:%d hops:%d \n",
+				action ? "PREP" : "PREQ", orig_addr, mgmt->sa, new_metric,
+				mpath->metric, hopcount);
 			if (mpath->flags & MESH_PATH_FIXED)
 				fresh_info = false;
-			else if ((mpath->flags & MESH_PATH_ACTIVE) &&
+			if ((mpath->flags & MESH_PATH_ACTIVE) &&
 			    (mpath->flags & MESH_PATH_SN_VALID)) {
 				if (SN_GT(mpath->sn, orig_sn) ||
-				    (mpath->sn == orig_sn &&
-				     (rcu_access_pointer(mpath->next_hop) !=
-						      sta ?
-					      mult_frac(new_metric, 10, 9) :
-					      new_metric) >= mpath->metric)) {
-					process = false;
-					fresh_info = false;
+					(mpath->metric &&
+					(rcu_access_pointer(mpath->next_hop) != sta) &&
+					(mult_frac(new_metric, 10, 9) >= mpath->metric))) {
+					 mhwmp_dbg(sdata, "skip long path orig:%pM sa:%pM \n",
+						orig_addr, mgmt->sa);
+					 process = false;
+					 fresh_info = false;
 				}
 			} else if (!(mpath->flags & MESH_PATH_ACTIVE)) {
 				bool have_sn, newer_sn, bounced;
 
+				mhwmp_dbg(sdata, "expired path for %pM flags %x \n", orig_addr, mpath->flags);
 				have_sn = mpath->flags & MESH_PATH_SN_VALID;
 				newer_sn = have_sn && SN_GT(orig_sn, mpath->sn);
 				bounced = have_sn &&
@@ -492,7 +524,13 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 					/* if SN is way different than what
 					 * we had then assume the other side
 					 * rebooted or restarted */;
+				} else if (have_sn && mpath->metric && (new_metric <= mpath->metric)) {
+					/* Accept frame */
 				} else {
+					mhwmp_dbg(sdata, "dropping %s from:%pM->%pM expired path"
+						"have_sn:%d new:%d bounce:%d delta:%d new_metric:%d metric:%d\n",
+						action ? "PREP" : "PREQ", orig_addr, mgmt->sa, have_sn, newer_sn,
+						bounced, SN_DELTA(orig_sn, mpath->sn), new_metric, mpath->metric);
 					process = false;
 					fresh_info = false;
 				}
@@ -504,6 +542,9 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 				return 0;
 			}
 			spin_lock_bh(&mpath->state_lock);
+			new_path = true;
+			mhwmp_dbg(sdata, "New Dst node path %pM with %s\n",
+			 orig_addr, action ? "PREP" : "PREQ");
 		}
 
 		if (fresh_info) {
@@ -515,6 +556,14 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 			mpath->flags |= MESH_PATH_SN_VALID;
 			mpath->metric = new_metric;
 			mpath->sn = orig_sn;
+			/* re-confirm if new paths are the optimal */
+			if (new_path) {
+				/* Overwrite expiry time with smaller value to re-confirm
+				 * if new path is optimal, with path refresh operation */
+				unsigned long refresh_time_jif =
+					msecs_to_jiffies(sdata->u.mesh.mshcfg.path_refresh_time);
+				exp_time = refresh_time_jif + jiffies;
+			}
 			mpath->exp_time = time_after(mpath->exp_time, exp_time)
 					  ?  mpath->exp_time : exp_time;
 			mpath->hop_count = hopcount;
@@ -525,6 +574,11 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 			ewma_mesh_fail_avg_init(&sta->mesh->fail_avg);
 			/* init it at a low value - 0 start is tricky */
 			ewma_mesh_fail_avg_add(&sta->mesh->fail_avg, 1);
+			mhwmp_dbg(sdata, "updated path %pM with %s\n",
+				mpath->dst, action ? "PREP" : "PREQ");
+#ifdef CONFIG_MAC80211_MHWMP_DEBUG
+			ieee80211_mpath_dump(sdata);
+#endif
 			mesh_path_tx_pending(mpath);
 			/* draft says preq_id should be saved to, but there does
 			 * not seem to be any use for it, skipping by now
@@ -534,8 +588,11 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 	}
 
 	/* Update and check transmitter routing info */
+	new_path = false;
+	exp_time = TU_TO_EXP_TIME(orig_lifetime);
 	ta = mgmt->sa;
-	if (ether_addr_equal(orig_addr, ta))
+	if (ether_addr_equal(orig_addr, ta) ||
+        (ifmsh->mshcfg.dot11MeshNolearn))
 		fresh_info = false;
 	else {
 		fresh_info = true;
@@ -550,11 +607,14 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 				       last_hop_metric) > mpath->metric)))
 				fresh_info = false;
 		} else {
+			new_path = true;
 			mpath = mesh_path_add(sdata, ta);
 			if (IS_ERR(mpath)) {
 				rcu_read_unlock();
 				return 0;
 			}
+			mhwmp_dbg(sdata, "New Tx node path %pM with %s\n",
+			 ta, action ? "PREP" : "PREQ");
 			spin_lock_bh(&mpath->state_lock);
 		}
 
@@ -565,6 +625,13 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 			}
 			mesh_path_assign_nexthop(mpath, sta);
 			mpath->metric = last_hop_metric;
+			if (new_path) {
+				/* Overwrite expiry time with smaller value to re-confirm
+				 * if new path is optimal, with path refresh operation */
+				unsigned long refresh_time_jif =
+					msecs_to_jiffies(sdata->u.mesh.mshcfg.path_refresh_time);
+				exp_time = refresh_time_jif + jiffies;
+			}
 			mpath->exp_time = time_after(mpath->exp_time, exp_time)
 					  ?  mpath->exp_time : exp_time;
 			mpath->hop_count = 1;
@@ -575,6 +642,11 @@ static u32 hwmp_route_info_get(struct ieee80211_sub_if_data *sdata,
 			ewma_mesh_fail_avg_init(&sta->mesh->fail_avg);
 			/* init it at a low value - 0 start is tricky */
 			ewma_mesh_fail_avg_add(&sta->mesh->fail_avg, 1);
+			mhwmp_dbg(sdata, "updated Tx node path %pM with %s\n",
+				ta, action ? "PREP" : "PREQ");
+#ifdef CONFIG_MAC80211_MHWMP_DEBUG
+			ieee80211_mpath_dump(sdata);
+#endif
 			mesh_path_tx_pending(mpath);
 		} else
 			spin_unlock_bh(&mpath->state_lock);
@@ -612,7 +684,7 @@ static void hwmp_preq_frame_process(struct ieee80211_sub_if_data *sdata,
 	mhwmp_dbg(sdata, "received PREQ from %pM\n", orig_addr);
 
 	if (ether_addr_equal(target_addr, sdata->vif.addr)) {
-		mhwmp_dbg(sdata, "PREQ is for us\n");
+		mhwmp_dbg(sdata, "PREQ is for us from %pM\n", mgmt->sa);
 		forward = false;
 		reply = true;
 		target_metric = 0;
@@ -817,6 +889,10 @@ static void hwmp_perr_frame_process(struct ieee80211_sub_if_data *sdata,
 		    !(mpath->flags & MESH_PATH_FIXED) &&
 		    (!(mpath->flags & MESH_PATH_SN_VALID) ||
 		    SN_GT(target_sn, mpath->sn)  || target_sn == 0)) {
+			mhwmp_dbg(sdata, "PERR deactivate path for :%pM flags:%x exp_time:%d"
+				"hops:%d metric:%d Queue_len:%d\n",
+				mpath->dst, mpath->flags, jiffies_to_msecs(mpath->exp_time - jiffies),
+				mpath->hop_count, mpath->metric, skb_queue_len(&mpath->frame_queue));
 			mpath->flags &= ~MESH_PATH_ACTIVE;
 			if (target_sn != 0)
 				mpath->sn = target_sn;
@@ -1102,6 +1178,7 @@ void mesh_path_start_discovery(struct ieee80211_sub_if_data *sdata)
 			mpath->flags |= MESH_PATH_RESOLVING;
 			mpath->discovery_retries = 0;
 			mpath->discovery_timeout = disc_timeout_jiff(sdata);
+			mhwmp_dbg(sdata, "New start discovery for %pM \n", mpath->dst);
 		}
 	} else if (!(mpath->flags & MESH_PATH_RESOLVING) ||
 			mpath->flags & MESH_PATH_RESOLVED) {
@@ -1182,6 +1259,7 @@ int mesh_nexthop_resolve(struct ieee80211_sub_if_data *sdata,
 	/* no nexthop found, start resolving */
 	mpath = mesh_path_lookup(sdata, target_addr);
 	if (!mpath) {
+		mhwmp_dbg(sdata, "no mpath found for %pM, add one \n", target_addr);
 		mpath = mesh_path_add(sdata, target_addr);
 		if (IS_ERR(mpath)) {
 			mesh_path_discard_frame(sdata, skb);
@@ -1189,13 +1267,16 @@ int mesh_nexthop_resolve(struct ieee80211_sub_if_data *sdata,
 		}
 	}
 
-	if (!(mpath->flags & MESH_PATH_RESOLVING) &&
-	    mesh_path_sel_is_hwmp(sdata))
-		mesh_queue_preq(mpath, PREQ_Q_F_START);
+	if (!(mpath->flags & (MESH_PATH_RESOLVING | MESH_PATH_REQ_QUEUED)) &&
+	    mesh_path_sel_is_hwmp(sdata)) {
+		mhwmp_dbg(sdata, "no nexthop found for %pM, start resolving \n", target_addr);
+		mesh_queue_preq(mpath, PREQ_Q_F_START | PREQ_Q_F_REFRESH);
+	}
 
-	if (skb_queue_len(&mpath->frame_queue) >= MESH_FRAME_QUEUE_LEN)
+	if (skb_queue_len(&mpath->frame_queue) >= MESH_FRAME_QUEUE_LEN) {
+        IEEE80211_IFSTA_MESH_CTR_INC(&sdata->u.mesh, dropped_frames_no_route);
 		skb_to_free = skb_dequeue(&mpath->frame_queue);
-
+	}
 	info->control.flags |= IEEE80211_TX_INTCFL_NEED_TXPROCESSING;
 	ieee80211_set_qos_hdr(sdata, skb);
 	skb_queue_tail(&mpath->frame_queue, skb);
@@ -1239,18 +1320,37 @@ static int mesh_nexthop_lookup_nolearn(struct ieee80211_sub_if_data *sdata,
 	return 0;
 }
 
+/**
+ * Checks if the given mpath is optimal.
+ *
+ * @param mpath: Pointer to the mpath structure containing path data.
+ * @return true if the mpath is optimal, false otherwise.
+ */
+bool hwmp_is_mpath_optimal(struct mesh_path *mpath)
+{
+    /* Check if the metric is within acceptable limits */
+    return (mpath->metric <= (mpath->hop_count * BEST_SINGLE_HOP_METRIC_HALOW));
+}
+
 void mesh_path_refresh(struct ieee80211_sub_if_data *sdata,
 		       struct mesh_path *mpath, const u8 *addr)
 {
-	if (mpath->flags & (MESH_PATH_REQ_QUEUED | MESH_PATH_FIXED |
-			    MESH_PATH_RESOLVING))
+	unsigned long refresh_time_jif = msecs_to_jiffies(sdata->u.mesh.mshcfg.path_refresh_time);
+
+	if ((mpath->flags & (MESH_PATH_REQ_QUEUED | MESH_PATH_FIXED | MESH_PATH_RESOLVING)))
 		return;
 
-	if (time_after(jiffies,
-		       mpath->exp_time -
-		       msecs_to_jiffies(sdata->u.mesh.mshcfg.path_refresh_time)) &&
-	    (!addr || ether_addr_equal(sdata->vif.addr, addr)))
+	if(time_after(jiffies, (mpath->exp_time - refresh_time_jif)) &&
+	   (!addr || ether_addr_equal(sdata->vif.addr, addr)))
 		mesh_queue_preq(mpath, PREQ_Q_F_START | PREQ_Q_F_REFRESH);
+
+	if (hwmp_is_mpath_optimal(mpath)) {
+		/* Refresh path, if path is being actively used & optimal */
+		unsigned long exp_time =
+			msecs_to_jiffies(sdata->u.mesh.mshcfg.dot11MeshHWMPactivePathTimeout);
+		mpath->exp_time = time_after(mpath->exp_time, (exp_time + jiffies)) ?
+			mpath->exp_time : (exp_time + jiffies);
+	}
 }
 
 /**
@@ -1311,10 +1411,11 @@ void mesh_path_timer(struct timer_list *t)
 		spin_unlock_bh(&mpath->state_lock);
 	} else if (mpath->discovery_retries < max_preq_retries(sdata)) {
 		++mpath->discovery_retries;
-		mpath->discovery_timeout *= 2;
+		mpath->discovery_timeout = min(mpath->discovery_timeout * 2,
+			msecs_to_jiffies(MESH_MAX_MPATH_DISCOVERY_TIMEOUT));
 		mpath->flags &= ~MESH_PATH_REQ_QUEUED;
 		spin_unlock_bh(&mpath->state_lock);
-		mesh_queue_preq(mpath, 0);
+		mesh_queue_preq(mpath, PREQ_Q_F_REFRESH);
 	} else {
 		mpath->flags &= ~(MESH_PATH_RESOLVING |
 				  MESH_PATH_RESOLVED |
diff --git a/net/mac80211/mesh_pathtbl.c b/net/mac80211/mesh_pathtbl.c
index 30c0d89203aff..5d4577bdbf3f8 100644
--- a/net/mac80211/mesh_pathtbl.c
+++ b/net/mac80211/mesh_pathtbl.c
@@ -779,6 +779,12 @@ void mesh_plink_broken(struct sta_info *sta)
 		if (rcu_access_pointer(mpath->next_hop) == sta &&
 		    mpath->flags & MESH_PATH_ACTIVE &&
 		    !(mpath->flags & MESH_PATH_FIXED)) {
+			mhwmp_dbg(sdata, "Broken path for :%pM flags:%x exp_time:%d"
+				"hops:%d metric:%d Queue_len:%d\n",
+				mpath->dst, mpath->flags,
+				jiffies_to_msecs(mpath->exp_time - jiffies),
+				mpath->hop_count, mpath->metric,
+				skb_queue_len(&mpath->frame_queue));
 			spin_lock_bh(&mpath->state_lock);
 			mpath->flags &= ~MESH_PATH_ACTIVE;
 			++mpath->sn;
diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 125496beecdd9..567a5f1caceb7 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -2978,19 +2978,13 @@ ieee80211_rx_mesh_data(struct ieee80211_sub_if_data *sdata, struct sta_info *sta
 		memcpy(fwd_hdr->addr2, sdata->vif.addr, ETH_ALEN);
 		/* update power mode indication when forwarding */
 		ieee80211_mps_set_frame_flags(sdata, NULL, fwd_hdr);
-	} else if (!mesh_nexthop_lookup(sdata, fwd_skb)) {
+	}
+	else if (!mesh_nexthop_resolve(sdata, fwd_skb)) {
 		/* mesh power mode flags updated in mesh_nexthop_lookup */
 		IEEE80211_IFSTA_MESH_CTR_INC(ifmsh, fwded_unicast);
 	} else {
-		/* unable to resolve next hop */
-		if (sta)
-			mesh_path_error_tx(sdata, ifmsh->mshcfg.element_ttl,
-					   hdr.addr3, 0,
-					   WLAN_REASON_MESH_PATH_NOFORWARD,
-					   sta->sta.addr);
-		IEEE80211_IFSTA_MESH_CTR_INC(ifmsh, dropped_frames_no_route);
-		kfree_skb(fwd_skb);
-		goto rx_accept;
+		/* skb queued: don't free */
+		return RX_DROP_MONITOR;
 	}
 
 	IEEE80211_IFSTA_MESH_CTR_INC(ifmsh, fwded_frames);
diff --git a/net/wireless/mesh.c b/net/wireless/mesh.c
index 2c6654075ca9e..fe8d42119a3c3 100644
--- a/net/wireless/mesh.c
+++ b/net/wireless/mesh.c
@@ -18,7 +18,7 @@
 #define MESH_CONF_T 		100
 #define MESH_HOLD_T 		100
 
-#define MESH_PATH_TIMEOUT	5000
+#define MESH_PATH_TIMEOUT	50000
 #define MESH_RANN_INTERVAL      5000
 #define MESH_PATH_TO_ROOT_TIMEOUT      6000
 #define MESH_ROOT_INTERVAL     5000
@@ -31,22 +31,22 @@
  */
 #define MESH_PREQ_MIN_INT	10
 #define MESH_PERR_MIN_INT	100
-#define MESH_DIAM_TRAVERSAL_TIME 50
+#define MESH_DIAM_TRAVERSAL_TIME 200
 
-#define MESH_RSSI_THRESHOLD	0
+#define MESH_RSSI_THRESHOLD	-85
 
 /*
  * A path will be refreshed if it is used PATH_REFRESH_TIME milliseconds
  * before timing out.  This way it will remain ACTIVE and no data frames
  * will be unnecessarily held in the pending queue.
  */
-#define MESH_PATH_REFRESH_TIME			1000
+#define MESH_PATH_REFRESH_TIME			10000
 #define MESH_MIN_DISCOVERY_TIMEOUT (2 * MESH_DIAM_TRAVERSAL_TIME)
 
 /* Default maximum number of established plinks per interface */
 #define MESH_MAX_ESTAB_PLINKS	32
 
-#define MESH_MAX_PREQ_RETRIES	4
+#define MESH_MAX_PREQ_RETRIES	6
 
 #define MESH_SYNC_NEIGHBOR_OFFSET_MAX 50
 
@@ -82,7 +82,7 @@ const struct mesh_config default_mesh_config = {
 	.power_mode = NL80211_MESH_POWER_ACTIVE,
 	.dot11MeshAwakeWindowDuration = MESH_DEFAULT_AWAKE_WINDOW,
 	.plink_timeout = MESH_DEFAULT_PLINK_TIMEOUT,
-	.dot11MeshNolearn = false,
+	.dot11MeshNolearn = true,
 };
 
 const struct mesh_setup default_mesh_setup = {
